clients/client-core/src/client/real_messages_control/real_traffic_stream.rs

```diff
@@ -222,8 +241,199 @@ where
     /// Accessor to the common instance of network topology.
     topology_access: TopologyAccessor,
 
-    /// Buffer containing all real messages received. It is first exhausted before more are pulled.
-    received_buffer: VecDeque<RealMessage>,
+    /// Buffer containing all real messages keyed by transmission lane.
+    received_buffer: ReceivedBuffer,
+
+    /// Incoming channel for being notified of closed connections, to avoid sending traffic
+    /// unnecessary
+    closed_connection_rx: ClosedConnectionReceiver,
+}
+
+#[derive(Default)]
+struct ReceivedBuffer {
+    buffer: HashMap<TransmissionLane, LaneBufferEntry>,
+}
+
+impl ReceivedBuffer {
+    fn is_empty(&self) -> bool {
+        self.buffer.is_empty()
+    }
+
+    fn remove(&mut self, lane: &TransmissionLane) -> Option<LaneBufferEntry> {
+        self.buffer.remove(lane)
+    }
+
+    fn num_lanes(&self) -> usize {
+        self.buffer.keys().count()
+    }
+
+    fn connections(&self) -> HashSet<u64> {
+        self.buffer
+            .keys()
+            .filter_map(|lane| match lane {
+                TransmissionLane::ConnectionId(id) => Some(id),
+                _ => None,
+            })
+            .copied()
+            .collect()
+    }
+
+    fn total_size(&self) -> usize {
+        self.buffer.values().map(LaneBufferEntry::len).sum()
+    }
+
+    fn total_size_in_bytes(&self) -> usize {
+        self.buffer
+            .values()
+            .map(|lane_buffer_entry| {
+                lane_buffer_entry
+                    .real_messages
+                    .iter()
+                    .map(|real_message| real_message.mix_packet.sphinx_packet().len())
+                    .sum::<usize>()
+            })
+            .sum()
+    }
+
+    fn get_oldest_set(&self) -> Vec<TransmissionLane> {
+        let mut buffer: Vec<_> = self
+            .buffer
+            .iter()
+            .map(|(k, v)| (k, v.age_in_millis()))
+            .collect();
+        buffer.sort_by_key(|v| v.1);
+        buffer
+            .iter()
+            .rev()
+            .map(|(k, _)| *k)
+            .take(OLDEST_LANE_SET_SIZE)
+            .copied()
+            .collect()
+    }
+
+    fn store(&mut self, lane: &TransmissionLane, real_messages: Vec<RealMessage>) {
```

Comment:
nitpicking here, but `self.buffer.entry(...)` would probably be cleaner

Reply:
Unfortunately in this case with `real_messages` being moved and the invariant of no empty entries, the correponding code using `entry` with `.and_modify(..).or_insert(..)` doesn't work :(